### PR TITLE
WIP - Lazy list completion

### PR DESCRIPTION
**Pull request by mahkoh**
Hi,

consider the following schema:

```graphql
type Query {
    items: [Item]
}

type Item {
    # fields omitted
}
```

It can happen that the list is too large to fit into memory--in particular if
multiple queries are running concurrently.

The obvious solution is to use paging, however,

1. Paging is inconvenient for the client.
2. Having the client execute multiple queries to get the whole list will
   increase the server load and increase the total response time.
3. If the problematic list is actually contained within another list, then it
   might be impossible to create a paging scheme that is both efficient and not
   overly complex.

Therefore we would like to

1. Fetch up to N elements of the list
2. Complete these elements
3. Send the completed elements to the client
4. Repeat unless all elements of the list have been processed

The intention of this PR is to allow the server to

1. Suspend the completion of fields of list type
2. Resume completion of the fields on-demand and after the overall GraphQL
   execution has completed
3. Stream the completed list elements so that no more than necessary have to be
   kept in memory

## Suspending a list completion

At the moment, the object returned by a `DataFetcher` of a field of list type
must be either

* `null`
* an array
* an instance of `Iterable`

This PR introduced the new functional interface `LazyList<T>` with the following
method:

```java
void forAll(BiConsumer<Boolean, ? super T> action);
```

If the `DataFetcher` of a field of list type returns an instance of this
interface, the completion of the list is suspended and the completion step
immediately returns a `completedFuture` with `data` equal to an instance of
`LazyList<Object>`.

The `LazyList<T>` returned by by the `DataFetcher` should, when invoked, call
the `action` once for each element of the list. (The meaning of the `Boolean`
argument will be explained below.) The `LazyList<Object>` returned by the
completion step will, when invoked, call the `action` once for each completed
element of the list.

This implies that the `Map<String, Object>` returned by the overall GraphQL
execution contains elements of type `LazyList<Object>` at its leaves.

## Serializing the overall execution result

Since the `Map<String, Object>` contains elements of type `LazyList<Object>`,
the server can no longer use stock Jackson to serialize the result. Instead,
they have to use a custom serializer to serialize the `LazyList<Object>`
objects.

This serializer has to

1. Write the opening `[`
2. Invoke `forAll` and serialize the objects passed to it via the callback
   in order
3. Write the closing `]`

Such a custom Jackson serializer is not part of this PR but it would not be too
hard to add, for example, a new `graphql-java-jackson-module` project that
contains such a serializer.

## Resuming the list completion

Once the `forAll` method of a `LazyList<Object>` is invoked, the completion of
the list resumes. This is quite similar to the usual list completion except that
we do not wait for the whole list to complete. Instead we pass the individual
elements of the list to the callback as soon as possible.

The server might not want completion to block on each element. Instead they
might want to complete, say, 10 elements concurrently and only then pass them to
the serializer. This is possible via the `Boolean` argument of the callback.

If the `Boolean` argument is `false`, then completion of the element will be
initiated but we don't wait for it to finish. Instead we add the
`CompletableFuture` to a list and proceed with the next element. Once the
`Boolean` argument becomes `true`, we wait for all `CompletableFutures` in the
list to complete and pass them on.

In the simplest case, the `Boolean` argument is always `true` and each element
is completed and serialized before we try to complete the next one.

## Error handling

Since the overall GraphQL execution completes before all fields have been
completed, the list of errors known at that point might not be complete.

Usually, the server is advised to call `toSpecification` on the overall
execution result to create a `Map<String, Object>` that is suitable for
serialization. However, since, in this map, the `error` element is not set if
there are no errors after the overall GraphQL execution completes, it is not
suitable for serializing trees that contain `LazyLists`.

Therefore, this PR introduces a new `ExecutionResult` implementation called
`LazyExecutionResult`. If the `Execution` class detects that `LazyLists` were
used during the execution, it wraps the overall execution result in a
`LazyExecutionResult`.

The `LazyExecutionResult` forwards the method `getData` to the wrapped delegate.
On the other hand, the `getErrors` method retrieves the current list of errors
from the `ExecutionContext` and `getExtensions` does something similar. The
`toSpecification` method is not supported if the server chooses to use
`LazyLists` and therefore throws an `UnsupportedOperationException`.

In order to serialize such an execution result, the server must

1. Write the opening `{`
2. Serialize the data (if any)
3. Serialize the errors (in the correct format) (if any at this point)
4. Serialize the extensions (if any)
5. Write the closing `}`

A custom Jackson serializer that knows how to serialize `ExecutionResult`
objects correctly would make the `toSpecification` method unnecessary (for
Jackson users) and could be added to a potential `graphql-java-jackson-module`
project.

## Compatibility with deferred results

In order to create compatibility with the @defer directive, I've added the
ability for fields to detect the cancellation of any of their parent fields.
For example, consider the following schema:

```graphql
type Query {
    value: Value
}

type Value {
    items: [Item]
    field: Int!
}
```

If `items` is a lazy list and the value fetched by `field` is null, then `value`
itself becomes null and `items` can never be completed. Therefore, it is now
possible for `items` to register a callback that is invoked when any parent
becomes null.

This is necessary because we want to start the deferred calls only after the
execution result of the overall GraphQL query has been fully completed. In order
to know when this is, we have to know when the number of pending lazy lists
becomes 0. This happens when each lazy list has either been fetched or
cancelled.

Since a lazy list might contain a @defer which itself might contain lazy
lists (and so on), the second commit in this series creates the necessary
framework to use both lazy completion and @defer in a single query.

## Incompatibility with the specification

`LazyLists` are incompatible with the GraphQL specification because we start
serializing the list before we know whether all elements can be retrieved and
completed. There are at least three cases that can cause this to fail:

1. The `forAll` method of the `LazyList<T>` throws an exception
2. The `forAll` method of the `LazyList<T>` passes a `null` element and the
   element type is non-null.
3. The `forAll` method of the `LazyList<T>` passes an element for which
   completion fails and the element type is non-null.

In these cases the whole list must be null according to the spec.

### CODE REVIEW

**Comment by mahkoh** [comment]
This PR is based on the last release and therefore has conflicts.

I've encountered the need for lazy lists in production and have already implemented much simpler support for lazy lists via a custom execution strategy. Such a toy implementation is, however, incompatible with deferred results and probably other features. Therefore I'd like to upstream support for lazy lists if possible.

I'm also not sure if this implementation does not break any other features that I'm not aware of. In particular, since the code that implements lazy lists is almost completely orthogonal to the existing code, the existing tests do not test the combination of this new feature and the existing features. I've added a test that tests deferred fields in combination with lazy lists.

**Comment by mahkoh** [comment]
To summarize what users have to do to use this feature:

1. Add a custom Jackson serializer for LazyList objects
2. Add a custom Jackson serializer for ExecutionResult objects
3. Return a LazyList object from the DataFetcher

Everything else is handled transparently.

**Comment by mahkoh** [comment]
I haven't written any documentation yet and I'd probably have to write more tests that test the interaction between this and other features (though I don't know which features are most likely to interact badly with lazy completion.)

But before I do this: Could you tell me if you're interested in merging support for lazy completion?

**Comment by bbakerman** [comment]
Sorry for the delay in getting back to you.  A few of us have been away with work.

A quick glance shows this to be a very complete PR and the code looks interesting.

I will look at it in more details and give you my thoughts on it

**Comment by bbakerman** [comment]
Once again thank you for this PR.

I havent fully got my head around the LazyList consumption but I will outline what I understand so far.

The results now support a LazyList which is really a callback for a list of results.  The consumer of that lazyList will be called back
one at a time as they are available until the "boolean true" is sent and then all results will be sent.  I am right here?

There is also a CompletionCancellationRegistry of callbacks but I could not see where anything was wired into that to get a calback.  I am
not clear on how this is intended to be used.

Squinting at it I feel like its a home grown "reactive stream" (http://www.reactive-streams.org/), where a LazyList is really a `org.reactivestreams.Publisher` 
and the callback actions are like the `org.reactivestreams.Subscriber` of those results.

Reactive streams allow you to have a list of results that are not all in memory at the one time and allows the consumer (the subscriber) to
say how many they want and also allows for a complete drain by having them ask for MAX results via `org.reactivestreams.Subscription#request`

Am I off the mark here in these thoughts?

If we were to have such a capability in graphql-java then I would model it using a more industry standard set of interfaces such
as reactive streams so it can be more familar to programmers.

In fact we modelled graphql subscriptions ` subscription { field { subfield1 }}` by using reactive streams and hence you get a stream
of results based on the top level fields.


The graphql specification (as it stands in 2018) makes assumptions that all results can be sent at the one time and that errors are all availave
which pulls the  implementation away from a "lazy" approach.  For example it mandates that IF a child field is null (but its type is non nullable)
then the parent field MUSt become nullable.  This is near impossible to do in a lazy list manner I suggest since you have already
leaked out parent results before the child completed (I think)

We have thought about having a reactive streams interface for graphql-java however one of the challenges of it (as you have discovered)
is the viral nature of how it changes the execution strategy implementation, right out to the type of results you get and how errors
are represented.

Some other people have created a execution strategy that uses reactive streams (https://github.com/bsideup/graphql-java-reactive/blob/master/core/src/main/java/com/github/bsideup/graphql/reactive/ReactiveExecutionStrategy.java) but they did so in isolation (eg copied most of Execution Strategy)
because its near impossible to have a "composeable" ExecutionStrategy based on CompleteableFutures (todays code) that can be composed
into a version that wants to have lazy streams of data (your example)

So to summarize I think in order to have lazy results in graphql-java I would see it implemented as follows

* as a reative stream shape
* as a seperate execution strategy
